packages/eslint-plugin/src/rules/no-wrapper-object-types.ts

```diff
@@ -0,0 +1,82 @@
+import type { TSESLint, TSESTree } from '@typescript-eslint/utils';
+import { AST_NODE_TYPES } from '@typescript-eslint/utils';
+
+import { createRule } from '../util';
+
+const classNames = new Set([
+  'BigInt',
+  // eslint-disable-next-line @typescript-eslint/internal/prefer-ast-types-enum
+  'Boolean',
+  'Function',
+  'Number',
+  'Object',
+  // eslint-disable-next-line @typescript-eslint/internal/prefer-ast-types-enum
+  'String',
+  'Symbol',
+]);
+
+export default createRule({
+  name: 'no-wrapper-object-types',
+  meta: {
+    type: 'problem',
+    docs: {
+      description: 'Disallow using confusing built-in primitive class wrappers',
+      recommended: 'recommended',
+    },
+    fixable: 'code',
+    messages: {
+      bannedClassType:
+        'Prefer using the primitive `{{preferred}}` as a type name, rather than the class wrapper `{{typeName}}`.',
```

Comment:
```suggestion
        'Prefer using the primitive `{{preferred}}` as a type name, rather than the upper-cased `{{typeName}}`.',
```

Opinion: These types aren't really class wrappers. They are interfaces that are technically a super-type of the primitive.

I think we should be less specific with the message here.